File: backend/routes/payment.js
import express from 'express';
import jwt from 'jsonwebtoken';

const createPaymentRouter = (stripe) => {
  const router = express.Router();

  // Middleware d'authentification local (même que cart.js)
  const auth = async (req, res, next) => {
    try {
      console.log('Vérification du token d\'authentification (payment)');
      const token = req.headers.authorization?.split(' ')[1];
      console.log('Token reçu:', token ? '<présent>' : '<absent>');

      if (!token) {
        console.log('Aucun token fourni');
        return res.status(401).json({ message: 'Authentification requise' });
      }

      // SÉCURITÉ: Vérifier que JWT_SECRET est défini
      if (!process.env.JWT_SECRET) {
        console.error('JWT_SECRET non défini dans payment.js');
        return res.status(500).json({ message: 'Erreur de configuration serveur' });
      }

      const decodedToken = jwt.verify(token, process.env.JWT_SECRET);
      console.log('Token décodé:', decodedToken);

      req.user = { id: decodedToken.id };
      next();
    } catch (error) {
      console.error('Erreur d\'authentification:', error);
      res.status(401).json({ message: 'Token invalide', error: error.message });
    }
  };

  // Middleware pour vérifier l'authentification
  router.use(auth);

  // Route pour créer une intention de paiement
  router.post('/create-payment-intent', async (req, res) => {
    try {
      const { amount, currency = 'eur', shippingInfo, shippingCost } = req.body;
      const userId = req.user.id; // Récupéré du middleware d'authentification

      // Préparer les metadata avec les informations de livraison
      const metadata = {
        userId: userId
      };

      if (shippingInfo) {
        metadata.firstName = shippingInfo.firstName || '';
        metadata.lastName = shippingInfo.lastName || '';
        metadata.email = shippingInfo.email || '';
        metadata.phone = shippingInfo.phone || '';
        metadata.address = shippingInfo.address || '';
        metadata.city = shippingInfo.city || '';
        metadata.postalCode = shippingInfo.postalCode || '';
        metadata.country = shippingInfo.country || 'France';
      }

      if (shippingCost) {
        metadata.shippingCost = shippingCost.toString();
      }

      // Créer l'intention de paiement avec support des méthodes modernes (Google Pay, Apple Pay, etc.)
      const paymentIntent = await stripe.paymentIntents.create({
        amount,
        currency,
        metadata,
        automatic_payment_methods: {
          enabled: true,
          allow_redirects: 'always',
        },
      });

      res.json({
        clientSecret: paymentIntent.client_secret
      });
    } catch (error) {
      console.error("Erreur lors de la création de l'intention de paiement:", error);
      res.status(500).json({ 
        message: "Erreur lors de la création de l'intention de paiement",
        error: error.message 
      });
    }
  });

  // Route pour confirmer le paiement
  router.post('/confirm-payment', async (req, res) => {
    try {
      const { paymentIntentId } = req.body;
      const userId = req.user.id;

      const paymentIntent = await stripe.paymentIntents.retrieve(paymentIntentId);

      if (paymentIntent.metadata.userId !== userId) {
        return res.status(403).json({ message: 'Non autorisé' });
      }

      res.json({ status: paymentIntent.status });
    } catch (error) {
      console.error('Erreur lors de la confirmation du paiement:', error);
      res.status(500).json({ 
        message: 'Erreur lors de la confirmation du paiement',
        error: error.message 
      });
    }
  });

  return router;
};

export default createPaymentRouter; 

File: Backend/routes/payment.py
from fastapi import APIRouter, HTTPException, Depends, status
from fastapi.security import HTTPBearer
from sqlalchemy.orm import Session
from pydantic import BaseModel
import stripe
import os
from database import get_db
from middleware.auth import verify_token
from models import User

router = APIRouter()
security = HTTPBearer()
stripe.api_key = os.getenv("STRIPE_SECRET_KEY")

class CreatePaymentIntentRequest(BaseModel):
    amount: int
    currency: str = "eur"
    shippingInfo: dict = None
    shippingCost: float = 0.0

class ConfirmPaymentRequest(BaseModel):
    paymentIntentId: str

@router.post("/create-payment-intent")
async def create_payment_intent(
    request: CreatePaymentIntentRequest,
    user: User = Depends(verify_token),
    db: Session = Depends(get_db)
):
    metadata = {"userId": str(user.id)}
    
    if request.shippingInfo:
        metadata.update({
            "firstName": request.shippingInfo.get("firstName", ""),
            "lastName": request.shippingInfo.get("lastName", ""),
            "email": request.shippingInfo.get("email", ""),
            "phone": request.shippingInfo.get("phone", ""),
            "address": request.shippingInfo.get("address", ""),
            "city": request.shippingInfo.get("city", ""),
            "postalCode": request.shippingInfo.get("postalCode", ""),
            "country": request.shippingInfo.get("country", "France")
        })
    
    if request.shippingCost:
        metadata["shippingCost"] = str(request.shippingCost)
    
    try:
        payment_intent = stripe.PaymentIntent.create(
            amount=request.amount,
            currency=request.currency,
            metadata=metadata,
            automatic_payment_methods={
                "enabled": True,
                "allow_redirects": "always"
            }
        )
        
        return {"clientSecret": payment_intent.client_secret}
    except stripe.error.StripeError as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Erreur lors de la création de l'intention de paiement: {str(e)}"
        )

@router.post("/confirm-payment")
async def confirm_payment(
    request: ConfirmPaymentRequest,
    user: User = Depends(verify_token),
    db: Session = Depends(get_db)
):
    try:
        payment_intent = stripe.PaymentIntent.retrieve(request.paymentIntentId)
        
        if payment_intent.metadata.get("userId") != str(user.id):
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Non autorisé"
            )
        
        return {"status": payment_intent.status}
    except stripe.error.StripeError as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Erreur lors de la confirmation du paiement: {str(e)}"
        )

